File: jsllab.c
#include "./src/llab.h"
#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#endif

void EMSCRIPTEN_KEEPALIVE wasm_string_null(char** str){
    (*str) = NULL;
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_string_allocation(char** str, int size){
    (*str) = (char*)malloc(sizeof(char)*size);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_insert_character_in_str(char** str, int size, int index, char c){
    if(index >= size)
        return;    
    (*str)[index] = c;
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_string_insert_terminator(char** str, int size, int index){
    if(index >= size)
        return;
    (*str)[index] = '\0';
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_internal_free(void** s){
    free(*s);
    return;
}

char EMSCRIPTEN_KEEPALIVE wasm_get_char_from_str(char** str, int index){
    return (*str)[index];
}

void EMSCRIPTEN_KEEPALIVE wasm_init_float_array(float** a, int size){
    (*a) = (float*)malloc(sizeof(float)*size);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_init_int_array(int** a, int size){
    (*a) = (int*)malloc(sizeof(int)*size);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_fill_float_array(float** a, int index, float v){
    (*a)[index] = v;
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_fill_int_array(int** a, int index, int v){
    (*a)[index] = v;
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_init_char_matrix(char*** c, int rows, int cols){
    (*c) = (char**)malloc(sizeof(char*)*rows);
    int i;
    for(i = 0; i < rows; i++){
        (*c)[i] = (char*)malloc(sizeof(char)*cols);
    }
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_fill_char_matrix(char*** c, int row, int col, char ch){
    (*c)[row][col] = ch;
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_init_float_matrix(float*** c, int rows, int cols){
    (*c) = (float**)malloc(sizeof(float*)*rows);
    int i;
    for(i = 0; i < rows; i++){
        (*c)[i] = (float*)malloc(sizeof(float)*cols);
    }
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_fill_float_matrix(float*** c, int row, int col, float v){
    (*c)[row][col] = v;
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_init_int_matrix(int*** c, int rows, int cols){
    (*c) = (int**)malloc(sizeof(int*)*rows);
    int i;
    for(i = 0; i < rows; i++){
        (*c)[i] = (int*)malloc(sizeof(int)*cols);
    }
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_fill_int_matrix(int*** c, int row, int col, int v){
    (*c)[row][col] = v;
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_fill_terminator_matrix(char*** c, int row, int col){
    (*c)[row][col] = '\0';
    return;
}

int EMSCRIPTEN_KEEPALIVE wasm_get_int_value_from_array(int** array, int index){
    return (*array)[index];
}

float EMSCRIPTEN_KEEPALIVE wasm_get_float_value_from_array(float** array, int index){
    return (*array)[index];
}

void EMSCRIPTEN_KEEPALIVE wasm_free_matrix(void*** m, int rows){
    int i;
    for(i = 0; i < rows; i++){
        free((*m)[i]);
    }
    free(*m);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_free_array(void** m){
    free(*m);
    return;
}   

void EMSCRIPTEN_KEEPALIVE wasm_load_model(model** m, char** filename){
    (*m) = load_model(*filename);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_free_model(model** m){
    free_model(*m);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_copy_model_without_learning_parameters(model** m, model** c){
    (*c) = copy_model_without_learning_parameters(*m);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_copy_multiple_model_without_learning_parameters(model** m, model*** c, int size){
    int i;
    *c = (model***)malloc(sizeof(model**)*size);
    for(i = 0; i < size; i++){
        (*c)[i] = copy_model_without_learning_parameters(*m);
    }
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_free_multiple_model_without_learning_parameters(model*** c, int size){
    int i;
    for(i = 0; i < size; i++){
        free_model_without_learning_parameters((*c)[i]);
    }
    free(*c);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_reset_model(model** m){
    reset_model(*m);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_reset_model_without_learning_parameters(model** m){
    reset_model_without_learning_parameters(*m);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_feed_forward(model** m, int tensor_depth, int tensor_i, int tensor_j, float** inputs){
	model_tensor_input_ff((*m),tensor_depth,tensor_i,tensor_j,(*inputs));
}

void EMSCRIPTEN_KEEPALIVE wasm_feed_forward_witout_learning_parameters(model** m, model** m_wlp,  int tensor_depth, int tensor_i, int tensor_j, float** inputs){
    model_tensor_input_ff_without_learning_parameters((*m_wlp),(*m),tensor_depth,tensor_i,tensor_j,(*inputs));
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_reset_multiple_model_without_learning_parameters(model*** m, int size){
    int i;
    for(i = 0; i < size; i++){
        reset_model_without_learning_parameters((*m)[i]);
    }
    return;
}

int EMSCRIPTEN_KEEPALIVE wasm_get_output_dimension_from_model(model** m){
    return get_output_dimension_from_model(*m);
}

void EMSCRIPTEN_KEEPALIVE wasm_feed_forward_opt_multi_thread(model** m , model*** ms, int number_of_threads, int tensor_depth, int tensor_i, int tensor_j,float*** inputs){
    model_tensor_input_ff_multicore_opt(*ms, *m, tensor_depth, tensor_i, tensor_j, *inputs, number_of_threads, number_of_threads);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_load_dueling_categorical_dqn(dueling_categorical_dqn** dqn, char** filename){
    *dqn = load_dueling_categorical_dqn(*filename);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_free_dueling_categorical_dqn(dueling_categorical_dqn** dqn){
    free_dueling_categorical_dqn(*dqn);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_free_dueling_categorical_dqn_without_learning_parameters(dueling_categorical_dqn** dqn){
    free_dueling_categorical_dqn_without_learning_parameters(*dqn);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_free_multiple_dueling_categorical_dqn_without_learning_parameters(dueling_categorical_dqn*** dqn, int size){
    int i;
    for(i = 0; i < size; i++){
        free_dueling_categorical_dqn_without_learning_parameters((*dqn)[i]);
    }
    free(*dqn);
    return;
}

int EMSCRIPTEN_KEEPALIVE wasm_get_dqn_input_size(dueling_categorical_dqn** dqn){
    return get_input_layer_size_dueling_categorical_dqn(*dqn);
}

void EMSCRIPTEN_KEEPALIVE wasm_copy_dueling_categorical_dqn_without_learning_parameters(dueling_categorical_dqn** dqn, dueling_categorical_dqn** copy){
    *copy = copy_dueling_categorical_dqn_without_learning_parameters(*dqn);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_copy_multiple_dueling_categorical_dqn_without_learning_parameters(dueling_categorical_dqn** dqn, dueling_categorical_dqn*** copy, int size){
    int i;
    *copy = (dueling_categorical_dqn***)malloc(sizeof(dueling_categorical_dqn**)*size);
    for(i = 0; i < size; i++){
        (*copy)[i] = copy_dueling_categorical_dqn_without_learning_parameters(*dqn);
    }
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_reset_dueling_categorical_dqn(dueling_categorical_dqn** dqn){
    reset_dueling_categorical_dqn(*dqn);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_reset_dueling_categorical_dqn_without_learning_parameters(dueling_categorical_dqn** dqn){
    reset_dueling_categorical_dqn_without_learning_parameters(*dqn);
    return;
}

void EMSCRIPTEN_KEEPALIVE wasm_reset_multiple_dueling_categorical_dqn_without_learning_parameters(dueling_categorical_dqn*** dqn, int size){
    int i;
    for(i = 0; i < size; i++){
        reset_dueling_categorical_dqn_without_learning_parameters((*dqn)[i]);
    }
    return;
}

float EMSCRIPTEN_KEEPALIVE wasm_get_float_value_from_model_output(model** m, int index){
    return get_output_layer_from_model(*m)[index];
}

float EMSCRIPTEN_KEEPALIVE wasm_get_float_value_from_models_output(model*** m, int i, int index){
    return get_output_layer_from_model((*m)[i])[index];
}

void EMSCRIPTEN_KEEPALIVE wasm_compute_q_function(dueling_categorical_dqn** dqn, float** input, float** output, int input_size, int output_size){
    int i;
    compute_probability_distribution(*input, input_size, *dqn);
    float* q = compute_q_functions(*dqn);
    for(i = 0; i < output_size; i++){
        (*output)[i] = q[i];
    }
}

int EMSCRIPTEN_KEEPALIVE wasm_get_best_action_from_float(float** input, int input_size){
    int i, index = 0;
    float max = (*input)[0];
    for(i = 0; i < input_size; i++){
        if((*input)[i] > max){
            max = (*input)[i];
            index = i;
        }
    }
    return index;
}

int EMSCRIPTEN_KEEPALIVE wasm_get_best_action_from_dqn(dueling_categorical_dqn** dqn, float** input, float** output, int input_size, int output_size){
    int i;
    compute_probability_distribution(*input, input_size, *dqn);
    float* q = compute_q_functions(*dqn);
    int index = 0;
    float max = q[0];
    for(i = 0; i < output_size; i++){
        if(q[i] > max){
            max = q[i];
            index = i;
        }
    }
    return index;
}

void EMSCRIPTEN_KEEPALIVE wasm_load_genome(genome** g, char** filename, int input_size, int output_size){
    (*g) = load_genome_complete(*filename);
}

void EMSCRIPTEN_KEEPALIVE wasm_free_genome(genome** g){
    free_genome(*g, get_global_innovation_number_connections_from_genome(*g));
}

void EMSCRIPTEN_KEEPALIVE wasm_ff_genome(genome** g, float** inputs, float** output){
    *output = feed_forward(*g, *inputs,get_global_innovation_number_nodes_from_genome(*g), get_global_innovation_number_connections_from_genome(*g));
}

File: jsllab.js
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

class Model{
    constructor(filename){
        
        this.filename = filename;
        this._filename = _malloc(4);
        
        _wasm_string_allocation(this._filename, this.filename.length+1);
        for(var i = 0; i < this.filename.length; i++){
            _wasm_insert_character_in_str(this._filename, this.filename.length, i, new TextEncoder().encode(this.filename)[i]);
        }
        _wasm_string_insert_terminator(this._filename, this.filename.length+1, i);
        
        this._model = _malloc(4);
        
        try{
            _wasm_load_model(this._model, this._filename);
        }
        catch{
            _wasm_free_array(this._filename);
            _free(this._filename);
            _free(this._model);
        }
        this.number_of_models = 0;
        this._models = null;
        
    }
    
    free(){
        try{
            _wasm_free_model(this._model);
            if(number_of_models > 0){
                _wasm_free_multiple_model_without_learning_parameters(this._models,this.number_of_models);
                _free(this._models);
            }
            _wasm_free_array(this._filename);
            _free(this._filename);
            _free(this._model);
        }
        catch{
            if(this.number_of_models > 0){
                _free(this._models);
            }
            _wasm_free_array(this._filename);
            _free(this._filename);
            _free(this._model);
        }
    }
    
    make_multi_thread(number_of_threads){
        if(number_of_threads < 1 || this.number_of_models > 0)
            return;
        this._models = _malloc(4);
        _wasm_copy_multiple_model_without_learning_parameters(this._model, this._models, number_of_threads);
    }
    
    reset(){
        _wasm_reset_model(this._model);
        _wasm_reset_multiple_model_without_learning_parameters(this._models, this,number_of_threads);
    }
    
    single_feed_forward(inputs, tensor_depth, tensor_i, tensor_j){
        var c_inputs = _malloc(4);
        var length = tensor_depth*tensor_i*tensor_j;
        _wasm_init_float_array(c_inputs, length);
        for(var i = 0; i < length; i++){
            _wasm_fill_float_array(c_inputs, i, inputs[i]);
        }
        
        _wasm_feed_forward(this._model, tensor_depth, tensor_i, tensor_j, c_inputs);
        
        var output_dim =  _wasm_get_output_dimension_from_model(this._model);
        var output = [];
        for(var i = 0; i < output_dim; i++){
            output.push(_wasm_get_float_value_from_model_output(this._model, i));
        }
        
        _wasm_free_array(c_inputs);
        _free(c_inputs);
        return output;
    }
    
    multiple_feed_forward(inputs, tensor_depth, tensor_i, tensor_j){
        if (this.number_of_threads < 1){
            return null;
        }
        var c_inputs = _malloc(4);
        var length = tensor_depth*tensor_i*tensor_j;
        _wasm_init_float_matrix(c_inputs, this.number_of_threads, this._models, tensor_depth, tensor_i, tensor_j, c_inputs);
        
        for(var i = 0; i < this.number_of_threads; i++){
            for(var j = 0; j < length; j++){
                _wasm_fill_float_matrix(c_inputs, i, j, inputs[i][j]);
            }
        }
        
        _wasm_feed_forward_opt_multi_thread(this._model , this._models, this.number_of_threads, tensor_depth,tensor_i,tensor_j,c_inputs);
        
        var output_dim = _wasm_get_output_dimension_from_model(this._model);
        var output = [];
        for(var i = 0; i < this.number_of_threads; i++){
            var list = [];
            for(var j = 0; j < output_dimension; j++){
                list.push(_wasm_get_float_value_from_models_output(this._models, i, j));
            }
            output.push(list);
        }
        _wasm_free_matrix(c_inputs, this.number_of_threads);
        _free(c_inputs);
        return output;
        
    }
    
}



class DQN{
    constructor(filename){
        
        this.filename = filename;
        this._filename = _malloc(4);
        
        _wasm_string_allocation(this._filename, this.filename.length+1);
        for(var i = 0; i < this.filename.length; i++){
            _wasm_insert_character_in_str(this._filename, this.filename.length, i, new TextEncoder().encode(this.filename)[i]);
        }
        _wasm_string_insert_terminator(this._filename, this.filename.length+1, i);
        
        this._model = _malloc(4);
        
        try{
            _wasm_load_dueling_categorical_dqn(this._model, this._filename);
        }
        catch{
            _wasm_free_array(this._filename);
            _free(this._filename);
            _free(this._model);
        }
    }
    
    free(){
        try{
            _wasm_free_dueling_categorical_dqn(this._model);
            _wasm_free_array(this._filename);
            _free(this._filename);
            _free(this._model);
        }
        catch{
            _wasm_free_array(this._filename);
            _free(this._filename);
            _free(this._model);
        }
    }

    reset(){
        _wasm_reset_dueling_categorical_dqn(this._model);
    }
    
    q_function(inputs, input_size, output_size){
 
        var c_inputs = _malloc(4);
        var length = input_size;
        _wasm_init_float_array(c_inputs, length);
        for(var i = 0; i < length; i++){
            _wasm_fill_float_array(c_inputs, i, inputs[i]);
        }
        
        var c_outputs = _malloc(4);
        _wasm_init_float_array(c_outputs,output_size);
        
        _wasm_compute_q_function(this._model, c_inputs, c_outputs, input_size,output_size);
        
        var output_dim =  output_size;
        var output = [];
        for(var i = 0; i < output_dim; i++){
            output.push(_wasm_get_float_value_from_array(this.c_outputs, i));
        }
        
        _wasm_free_array(c_inputs);
        _wasm_free_array(c_outputs);
        _free(c_inputs);
        _free(c_outputs);
        return output;
    }
    
    get_best_action(q_function){
        var index = 0;
        var max = q_function[0];
        for(var i = 1; i < q_function.length; i++){
            if(q_function[i] > max){
                max = q_function[i];
                index = i;
            }
        }
        return index;
    }
}

class Genome{
    constructor(filename, input_size, output_size){
        
        this.filename = filename;
        this.input_size = input_size;
        this.output_size = output_size;
        this._filename = _malloc(4);
        
        _wasm_string_allocation(this._filename, this.filename.length+1);
        for(var i = 0; i < this.filename.length; i++){
            _wasm_insert_character_in_str(this._filename, this.filename.length, i, new TextEncoder().encode(this.filename)[i]);
        }
        _wasm_string_insert_terminator(this._filename, this.filename.length+1, i);
        
        this._model = _malloc(4);
        
        try{
            _wasm_load_genome(this._model, this._filename, input_size, output_size);
        }
        catch{
            _wasm_free_array(this._filename);
            _free(this._filename);
            _free(this._model);
        }
    }
    
    free(){
        try{
            _wasm_free_genome(this._model);
            _wasm_free_array(this._filename);
            _free(this._filename);
            _free(this._model);
        }
        catch{
            _wasm_free_array(this._filename);
            _free(this._filename);
            _free(this._model);
        }
    }

    ff(inputs){
        
        var c_inputs = _malloc(4);
        var c_outputs = _malloc(4);
        var length = this.input_size;
        _wasm_init_float_array(c_inputs, length);
        for(var i = 0; i < length; i++){
            _wasm_fill_float_array(c_inputs, i, inputs[i]);
        }
        _wasm_ff_genome(this._model, c_inputs,c_outputs);
        
        var output_dim =  this.output_size;
        var output = [];
        for(var i = 0; i < output_dim; i++){
            output.push(_wasm_get_float_value_from_array(c_outputs, i));
        }
        
        _wasm_free_array(c_inputs);
        _wasm_free_array(c_outputs);
        _free(c_inputs);
        _free(c_outputs);
        return output;
        
    }
}

